lib/Arch/PPC_C.cpp

```diff
@@ -0,0 +1,332 @@
+    RegisterConstraint({VariantConstraint("F3", kTypeFloat, kMaxBit32)}),
+    RegisterConstraint({VariantConstraint("F4", kTypeFloat, kMaxBit32)}),
+};
+
+// Used to split things like `i64`s into multiple `i32`s.
+static llvm::Type *IntegerTypeSplitter(llvm::Type *type) {
+  auto int_ty = llvm::dyn_cast<llvm::IntegerType>(type);
+  if (!int_ty) {
+    return nullptr;
+  }
+
+  auto width = int_ty->getPrimitiveSizeInBits();
+  if (width <= 32) {
+    return nullptr;
+  }
+
+  auto num_elements = (width + 31) / 32;
+  auto i32_ty = llvm::Type::getInt32Ty(type->getContext());
+  return llvm::ArrayType::get(i32_ty, num_elements);
+}
+
+}  // namespace
+
+class PPC_C : public CallingConvention {
+ public:
+  explicit PPC_C(const remill::Arch *arch);
+  virtual ~PPC_C() = default;
+
+  llvm::Error AllocateSignature(FunctionDecl &fdecl,
+                                llvm::Function &func) override;
+
+ private:
+  llvm::Error BindParameters(llvm::Function &function, bool injected_sret,
+                             std::vector<ParameterDecl> &param_decls);
+
+  llvm::Error BindReturnValues(llvm::Function &function, bool &injected_sret,
+                               std::vector<ValueDecl> &ret_decls);
+
+  const std::vector<RegisterConstraint> &parameter_register_constraints;
+  const std::vector<RegisterConstraint> &return_register_constraints;
+};
+
+std::unique_ptr<CallingConvention>
+CallingConvention::CreatePPC_C(const remill::Arch *arch) {
+  return std::make_unique<PPC_C>(arch);
+}
+
+PPC_C::PPC_C(const remill::Arch *arch)
+    : CallingConvention(llvm::CallingConv::C, arch),
+      parameter_register_constraints(kParamRegConstraints),
+      return_register_constraints(kReturnRegConstraints) {}
+
+llvm::Error PPC_C::AllocateSignature(FunctionDecl &fdecl,
+                                     llvm::Function &func) {
+
+  // Bind return values first to see if we have injected an sret into the
+  // parameter list. Then, bind the parameters. It is important that we bind the
+  // return values before the parameters in case we inject an sret.
+  bool injected_sret = false;
+  auto err = BindReturnValues(func, injected_sret, fdecl.returns);
+  if (remill::IsError(err)) {
+    return err;
+  }
+
+  err = BindParameters(func, injected_sret, fdecl.params);
+  if (remill::IsError(err)) {
+    return err;
+  }
+
+  fdecl.return_stack_pointer_offset = 0;
+  fdecl.return_stack_pointer = arch->RegisterByName("R1");
+
+  fdecl.return_address.reg = arch->RegisterByName("LR");
+  fdecl.return_address.type = fdecl.return_address.reg->type;
+
+  return llvm::Error::success();
+}
+
+llvm::Error
+PPC_C::BindReturnValues(llvm::Function &function, bool &injected_sret,
+                        std::vector<anvill::ValueDecl> &ret_values) {
+
+  llvm::Type *ret_type = function.getReturnType();
+  LOG(INFO) << "Binding on return " << remill::LLVMThingToString(ret_type);
+  injected_sret = false;
```

Comment:
Modifies argument even if result is failure, is this desired? Consider using tuple as return type (since bool is easily default constructible).

Reply:
[5e99152](https://github.com/lifting-bits/anvill/pull/340/commits/5e99152bc1fbc258ae6dcca43e0a96186137640b)